libssh-rs/src/lib.rs

```diff
@@ -1271,6 +1271,74 @@ impl SshKey {
             return Ok(SshKey { key });
         }
     }
+
+    /// Import a certificate from the given filename.
+    ///
+    /// # Arguments
+    /// * `filename` - The path to the certificate.
+    ///
+    /// # Errors
+    /// Will return [`Err`] if failed to parse ssh key from file.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use libssh_rs::SshKey;
+    /// let cert_path = "/path/to/my/key-cert.pub";
+    /// # let cert_path = "./test/test-key-cert.pub";
+    ///
+    /// let cert = SshKey::from_cert_file(cert_path).unwrap();
+    /// ```
+    pub fn from_cert_file(filename: &str) -> SshResult<SshKey> {
+        let filename_cstr = CString::new(filename).map_err(|e| {
+            Error::Fatal(format!(
+                "Could not make CString from filename '{filename}': {e:#}"
+            ))
+        })?;
+        unsafe {
+            let mut key = sys::ssh_key_new();
+            if sys::ssh_pki_import_cert_file(filename_cstr.as_ptr(), &mut key) != sys::SSH_OK as i32
+            {
+                sys::ssh_key_free(key);
+                return Err(Error::Fatal(format!(
+                    "Failed to parse ssh key from file '{filename}'"
+                )));
+            }
+            return Ok(SshKey { key });
+        }
+    }
+
+    /// Copy the certificate part of a public key into a private key.
+    ///
+    /// # Arguments
+    /// * `privkey` - Copy the certificate part of a public key into a private key.
+    ///
+    /// # Errors
+    /// Will return [`Err`] if failed to copy cert to privkey.
+    ///
+    /// # Example
+    /// ```rust
+    /// # use libssh_rs::SshKey;
+    /// let privkey_path = "/path/to/my/key";
+    /// let cert_path = "/path/to/my/key-cert.pub";
+    ///
+    /// # let privkey_path = "./test/test-key";
+    /// # let cert_path = "./test/test-key-cert.pub";
+    /// #
+    /// let privkey = SshKey::from_privkey_file(privkey_path, None).unwrap();
+    /// let cert = SshKey::from_cert_file(cert_path).unwrap();
+    ///
+    /// cert.copy_to(&privkey).unwrap();
+    ///
+    /// ```
+    pub fn copy_to(&self, privkey: &Self) -> SshResult<()> {
```

Comment:
I think this API (the underlying one from libssh) is confusingly named and a bit obscure in purpose, and that bubbles up here to also make this one confusing.

I think the main conceptual issue is that it isn't obvious which way the data is flowing, and because both private  and public keys have the same types, it's difficult to use the type system to catch mistakes.

I think there are a couple of levels of things that can be done to improve this.

1. Rename this method to something like `copy_certificate_to_private_key`
2. Make `privkey: &mut Self` to make it clearer that it is privkey that will be mutated by this method, and not `&self`.  Alternatively, does it make sense to invert this method and make it `assign_certificate(&mut self, pubkey: &Self)` and flow the other way around?
3. Is more involved: introduce distinct Private and Public key types with APIs that only allow appropriate underlying libssh functions to be called in the appropriate contexts.  That one is a lot more work!

I think the example successfully shows how to use this method, but it doesn't really explain why or when you would need this.  Do you think we could flesh that out?

Reply:
@DasKoebi I just noticed that this PR is still outstanding. Are you planning to follow up on this? Thanks!

Reply:
Sorry, I forgot about that as we ended up using a completely different solution.

I renamed the method and added the `mut` to the parameter as you suggested. As for the example: I'm not quite sure why exactly I needed this anymore 😅, but I think it's needed if the client's private key is signed by a certificate on the host. Since I'm not 100% sure about that, I didn't change the example.

I can't imagine putting much more effort into this. I would be happy if this gets merged, but I understand if you want to reject such a unpolished solution.

Reply:
if this did get merged, would your own code use it? If not, I'd be inclined to close the PR without merging, on the basis that less code is better and if no one is using it, then there's not a lot of point to keeping it around!

Reply:
We are currently using a completely different solution, so it's okay if you close the PR.

Reply:
Thanks for confirming!